server/src/main/java/org/opensearch/index/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,837 @@
+            }
+
+            return new SourceValueFetcher(name(), context, nullValue) {
+                @Override
+                protected String parseSourceValue(Object value) {
+                    String keywordValue = value.toString();
+                    if (keywordValue.length() > ignoreAbove) {
+                        return null;
+                    }
+
+                    NamedAnalyzer normalizer = normalizer();
+                    if (normalizer == null) {
+                        return keywordValue;
+                    }
+
+                    try {
+                        return normalizeValue(normalizer, name(), keywordValue);
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                }
+            };
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        NamedAnalyzer normalizer() {
+            return indexAnalyzer();
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, Supplier<SearchLookup> searchLookup) {
+            failIfNoDocValues();
+            return new SortedSetOrdinalsIndexFieldData.Builder(name(), CoreValuesSourceType.BYTES);
+        }
+
+        @Override
+        public Query fuzzyQuery(
+            Object value,
+            Fuzziness fuzziness,
+            int prefixLength,
+            int maxExpansions,
+            boolean transpositions,
+            QueryShardContext context
+        ) {
+            // TODO: Not sure if we can reasonably describe a fuzzy query in terms of n-grams without exploding the cardinality
+            throw new IllegalArgumentException(
+                "Can only use fuzzy queries on keyword and text fields - not on [" + name() + "] which is of type [" + typeName() + "]"
+            );
+        }
+
+        @Override
+        public Query prefixQuery(String value, MultiTermQuery.RewriteMethod method, boolean caseInsensitive, QueryShardContext context) {
+            return wildcardQuery(value + "*", method, caseInsensitive, context);
+        }
+
+        @Override
+        public Query wildcardQuery(String value, MultiTermQuery.RewriteMethod method, boolean caseInsensitive, QueryShardContext context) {
+            NamedAnalyzer normalizer = normalizer();
+            if (normalizer != null) {
+                value = normalizeWildcardPattern(name(), value, normalizer);
+            }
+            final String finalValue;
+            if (caseInsensitive) {
+                // Use ROOT locale, as it seems to be consistent with AutomatonQueries.toCaseInsensitiveChar.
+                finalValue = value.toLowerCase(Locale.ROOT);
+            } else {
+                finalValue = value;
+            }
+            Predicate<String> matchPredicate;
+            if (value.contains("?")) {
```

Comment:
Do we need to handle `*` as a special case of "field exists" here? (the test case below)

```
"wildcard query matches all":
  - do:
      search:
        index: test
        body:
          query:
            wildcard:
              my_field:
                value: "*"
```

Reply:
That's a good idea! I suspect that it's not working properly. Similarly, a regexp query for `.*` should do the same (rather than `MatchAllDocsQuery`).

Reply:
Curiously, the YAML REST test **does** pass, but it looks like it's producing a `MatchAllDocsQuery`, and then the `valueFetcher` fetches each value. I added an extra document that doesn't have the field, and it still works, so if the `valueFetcher` doesn't find something the `match` method must return false false.

Anyway, in case the wildcard field is sparse, we should do a field exists query as the approximation anyway, instead of `MatchAllDocs`. I'll add that change now.

Reply:
So -- I apparently wasn't running the YAML REST test previously 😞 🫴 . 

I had renamed the YAML test file at some point before committing it, so my test filter was matching nothing (and passing).

I just updated the commit with a real fix, along with more robust handling of (and unit tests for) the `*`, `.*`, `?`, and `.` cases.